pep-0684.rst

```diff
@@ -0,0 +1,640 @@
+extensions, particularly when they are doing the work in their spare
+time.  The concerns they have expressed are understandable.  We address
+the partial solution in `Restricting Extension Modules`_ below.
+
+Alternate Python Implementations
+--------------------------------
+
+Other Python implementation are not required to provide support for
+multiple interpreters in the same process (though some do already).
+
+Security Implications
+---------------------
+
+There is no known impact to security with this proposal.
+
+Maintainability
+---------------
+
+On the one hand, this proposal has already motivated a number of
+improvements that make CPython *more* maintainable.  That is expected
+to continue.  On the other hand, the underlying work has already
+exposed various pre-existing defects in the runtime that have had
+to be fixed.  That is also expected to continue as multiple interpreters
+receive more use.  Otherwise there shouldn't be a significant impact
+on maintainability.  The effect should be net positive.
+
+Performance
+-----------
+
+The work to consolidate globals has already provided a number of
+improvements to CPython's performance, both speeding it up and using
+less memory.  This should continue.  Performance benefits to a
+per-interpreter GIL have not been explored.  At the least it is
+not expected to make CPython slower (as long as interpreters are
+sufficiently isolated).
+
+
+Specification
+=============
+
+As summarized `above <High-Level Summary_>`_, this proposal involves the
+following changes, in the order they must happen:
+
+1. `consolidate global runtime state <Consolidating Runtime Global State_>`_
+   (including objects) into ``_PyRuntimeState``
+2. move nearly all of the state down into ``PyInterpreterState``
+3. finally, move the GIL down into ``PyInterpreterState``
+4. everything else
+
+   * add to the public C-API
+   * implement restrictions in ``ExtensionFileLoader``
+   * work with popular extension maintainers to help with multi-interpreter support
+
+Per-Interpreter State
+---------------------
+
+The following runtime state will be moved to ``PyInterpreterState``:
+
+* all global objects that are not safely shareable (fully immutable)
+* the GIL
+* mutable, currently protected by the GIL
+* mutable, currently protected by some other per-interpreter lock
+* mutable, may be used independently in different interpreters
+* all other mutable (or effectively mutable) state not otherwise excluded below
+
+The following state will not be moved:
+
+* global objects that are safely shareable, if any
+* immutable, often ``const``
+* treated as immutable
+* related to CPython's ``main()`` execution
+* related to the REPL
+* set during runtime init, then treated as immutable
+* mutable, protected by some global lock
+* mutable, atomic
+
+Also, a number of parts of the global state were already moved to the
+interpreter, such as GC, warnings, and atexit hooks.
```

Comment:
```suggestion
interpreter, such as GC, warnings, and atexit hooks, will also not be moved.
```

Clarify and avoid incomplete sentence

Reply:
I'll need to move this, since the context is confusing.